consensus/types/src/contribution_and_proof.rs

```diff
@@ -0,0 +1,82 @@
+use super::{
+    ChainSpec, Domain, EthSpec, Fork, Hash256, PublicKey, SecretKey, Signature, SignedRoot,
+    SyncCommitteeContribution, SyncSelectionProof,
+};
+use crate::test_utils::TestRandom;
+use serde_derive::{Deserialize, Serialize};
+use ssz_derive::{Decode, Encode};
+use test_random_derive::TestRandom;
+use tree_hash_derive::TreeHash;
+
+/// A Validators aggregate sync committee contribution and selection proof.
+///
+/// Spec v0.12.1
+#[cfg_attr(feature = "arbitrary-fuzz", derive(arbitrary::Arbitrary))]
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Encode, Decode, TestRandom, TreeHash)]
+#[serde(bound = "T: EthSpec")]
+pub struct ContributionAndProof<T: EthSpec> {
+    /// The index of the validator that created the aggregate contribution.
+    #[serde(with = "serde_utils::quoted_u64")]
+    pub aggregator_index: u64,
+    /// The aggregate attestation.
+    pub contribution: SyncCommitteeContribution<T>,
+    /// A proof provided by the validator that permits them to publish on the
+    /// `beacon_aggregate_and_proof` gossipsub topic.
+    pub selection_proof: Signature,
+}
+
+impl<T: EthSpec> ContributionAndProof<T> {
+    /// Produces a new `ContributionAndProof` with a `selection_proof` generated by signing
+    /// `aggregate.data.slot` with `secret_key`.
+    ///
+    /// If `selection_proof.is_none()` it will be computed locally.
+    pub fn from_aggregate(
+        aggregator_index: u64,
+        contribution: SyncCommitteeContribution<T>,
+        selection_proof: Option<SyncSelectionProof>,
+        secret_key: &SecretKey,
+        fork: &Fork,
+        genesis_validators_root: Hash256,
+        spec: &ChainSpec,
+    ) -> Self {
+        let selection_proof = selection_proof
+            .unwrap_or_else(|| {
+                SyncSelectionProof::new::<T>(
+                    contribution.slot,
+                    contribution.subcommittee_index,
+                    secret_key,
+                    fork,
+                    genesis_validators_root,
+                    spec,
+                )
+            })
+            .into();
+
+        Self {
+            aggregator_index,
+            contribution,
+            selection_proof,
+        }
+    }
+
+    /// Returns `true` if `validator_pubkey` signed over `contribution.slot`.
+    pub fn is_valid_selection_proof(
+        &self,
+        validator_pubkey: &PublicKey,
+        fork: &Fork,
+        genesis_validators_root: Hash256,
+        spec: &ChainSpec,
+    ) -> bool {
+        let target_epoch = self.contribution.slot.epoch(T::slots_per_epoch());
+        let domain = spec.get_domain(
+            target_epoch,
+            Domain::SyncCommitteeSelectionProof,
+            fork,
+            genesis_validators_root,
+        );
+        let message = self.contribution.slot.signing_root(domain);
```

Comment:
I noticed this was out of date and fixed it on my branch (https://github.com/michaelsproul/lighthouse/commit/4b2db537e7e3c147e6c2b5435bcf415fbf23210e) if you'd like to cherry-pick that commit.

At some point we can also decide whether we need multiple code paths for verification, perhaps the signature set one is sufficient

Reply:
Thanks! Yeah I haven't used the `is_valid` method anywhere I don't think. And I also don't see where the equivalent method in `SignedAggregateAndProof` is used. I'd be on board with getting rid of them